pkg/ovs/openflow/ofctrl_packetin.go

```diff
@@ -40,15 +51,41 @@ func GetTCPHeaderData(ipPkt util.Message) (tcpSrcPort, tcpDstPort uint16, tcpSeq
 		tcpBytes, err = typedIPPkt.Data.(*util.Buffer).MarshalBinary()
 	}
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
-	tcpIn := new(protocol.TCP)
-	err = tcpIn.UnmarshalBinary(tcpBytes)
+	tcpPkt = new(protocol.TCP)
+	err = tcpPkt.UnmarshalBinary(tcpBytes)
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
 
-	return tcpIn.PortSrc, tcpIn.PortDst, tcpIn.SeqNum, tcpIn.AckNum, tcpIn.Code, nil
+	return tcpPkt, nil
+}
+
+func GetTCPDataWithoutOptions(tcpPkt *protocol.TCP) (data []byte, err error) {
+	// TCP.HdrLen is 4-octet unit indicating the length of TCP header including options.
+	tcpOptionsLen := (tcpPkt.HdrLen - tcpStandardHdrLen) * 4
+	// Trim TCP option end mark. From RFC 793:
+	//   Specific Option Definitions
+	//     End of Option List
+	//        +--------+
+	//        |00000000|
+	//        +--------+
+	//         Kind=0
+	//
+	//        This option code indicates the end of the option list.  This
+	//        might not coincide with the end of the TCP header according to
+	//        the Data Offset field.  This is used at the end of all options,
+	//        not the end of each option, and need only be used if the end of
+	//        the options would not otherwise coincide with the end of the TCP
+	//        header.
+	if tcpPkt.Data[tcpOptionsLen] == tcpOptionEndKind {
+		tcpOptionsLen += 2
+	}
```

Comment:
I feel https://www.rfc-editor.org/rfc/rfc793#section-3.1:~:text=Specific%20Option%20Definitions is misunderstood. I think "This might not coincide with the end of the TCP header according to the Data Offset field" doesn't mean the case "Options exceeding the Data Offset" but means the opposite, for example: when an option takes 5 bytes, but the data offset must +2 to accommodate 5 bytes, then option-kind 0 could be used to indicate the end of all options and two bytes of padding can fill in the rest.

```
+--------+--------+--------+--------+--------+--------+--------+--------+
|00000010|00000101|        option data       |00000000|00000000|00000000|
+--------+--------+--------+--------+--------+--------+--------+--------+
  kind=2   length            data              Kind=0       Padding
```  

I think code in linux kernel and gopacket can confirm it:
https://github.com/torvalds/linux/blob/6a98c9cae232800c319ed69e1063480d31430887/net/ipv4/tcp_input.c#L4042-L4054
https://github.com/google/gopacket/blob/32ee38206866f44a74a6033ec26aeeb474506804/layers/tcp.go#L271-L282

And note that the description of Padding also indicates data always begins on a 32 bit boundary:

```
  Padding:  variable

    The TCP header padding is used to ensure that the TCP header ends
    and data begins on a 32 bit boundary.  The padding is composed of
    zeros.
```

Reply:
Emm that's weird. If I don't move two more bytes the lib that we used to inspect DNS response couldn't work.

Reply:
That's application layer length field according to the RFC.
 
> DNS clients and servers SHOULD pass the two-octet length field, and
   the message described by that length field, to the TCP layer at the
   same time (e.g., in a single "write" system call) to make it more
   likely that all the data will be transmitted in a single TCP segment.

https://www.rfc-editor.org/rfc/rfc7766#page-11